pkg/controller/network/resourcerecordset/controller.go

```diff
@@ -0,0 +1,179 @@
+
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane/provider-aws/apis/network/v1alpha3"
+	"github.com/crossplane/provider-aws/pkg/clients/resourcerecordset"
+	"github.com/crossplane/provider-aws/pkg/clients/zone"
+	"github.com/crossplane/provider-aws/pkg/controller/utils"
+)
+
+const (
+	errUnexpectedObject = "The managed resource is not an ResourceRecordSet resource"
+	errList             = "failed to list the ResourceRecordSet resource"
+	errCreate           = "failed to create the ResourceRecordSet resource"
+	errUpdate           = "failed to update the ResourceRecordSet resource"
+	errDelete           = "failed to delete the ResourceRecordSet resource"
+	errState            = "failed to determine resource state"
+)
+
+// SetupResourceRecordSet adds a controller that reconciles ResourceRecordSets.
+func SetupResourceRecordSet(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha3.ResourceRecordSetGroupKind)
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha3.ResourceRecordSet{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha3.ResourceRecordSetGroupVersionKind),
+			managed.WithExternalConnecter(&connector{client: mgr.GetClient(), newClientFn: resourcerecordset.NewClient, awsConfigFn: utils.RetrieveAwsConfigFromProvider}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithConnectionPublishers(),
+			managed.WithInitializers(),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	client      client.Client
+	newClientFn func(*aws.Config) resourcerecordset.Client
+	awsConfigFn func(context.Context, client.Reader, *corev1.ObjectReference) (*aws.Config, error)
+}
+
+func (conn *connector) Connect(ctx context.Context, mgd resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mgd.(*v1alpha3.ResourceRecordSet)
+	if !ok {
+		return nil, errors.New(errUnexpectedObject)
+	}
+
+	awsconfig, err := conn.awsConfigFn(ctx, conn.client, cr.Spec.ProviderReference)
+	if err != nil {
+		return nil, err
+	}
+
+	c := conn.newClientFn(awsconfig)
+
+	return &external{kube: conn.client, client: c}, nil
+}
+
+type external struct {
+	kube   client.Client
+	client resourcerecordset.Client
+}
+
+func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mgd.(*v1alpha3.ResourceRecordSet)
+
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+
+	rrset, err := resourcerecordset.GetResourceRecordSet(ctx, e.client, cr.Spec.ForProvider)
+	if _, ok := err.(*resourcerecordset.NotFound); ok {
```

Comment:
We just implement a `func IsErrorNotFound(err error) bool` in clients folder for each resource if we can't use other checks and use that to check error type instead of doing the casting here.

Reply:
Suggest change has been incorporated.